reactive/etcd.py

```diff
@@ -326,8 +326,21 @@ def register_node_with_leader(cluster):
 
     for unit in peers:
         if 'client_urls' not in peers[unit].keys():
-            # we cannot register. State not attainable.
             msg = 'Waiting for unit to complete registration.'
+            if ('peer_urls' in peers[unit].keys() and
+                    peers[unit]['peer_urls'] and
+                    get_ingress_address('cluster') in peers[unit]['peer_urls'] and  # noqa
+                    not host.service_running(bag.etcd_daemon)):
+                # We have a peer that is unstarted and it is this node.
+                # We do not run etcd now. Instead of blocking everyone
+                # try to self-unregister.
+                try:
+                    leader_address = leader_get('leader_address')
+                    msg = 'Etcd service did not start. Will retry soon.'
+                    etcdctl.unregister(peers[unit]['unit_id'], leader_address)
+                except CalledProcessError:
+                    log('Notice:  Unit failed to unregister', 'WARNING')
+            # we cannot register. State not attainable.
             status_set('waiting', msg)
             return
```

Comment:
This looks to me like it will unregister, return, and retry on the next hook (up to 5 min delay). Why not continue on and retry registration immediately?

Reply:
Indeed this is the intend here, to retry on the next reactive cycle.

I was not able to reproduce the respective issue consistently. It happened only once. So I cannot tell you what would happen if we try to immediately register, I can only mitigate the situation I saw by unregistering the node and waiting.

There are two main reasons why I would not go for the immediately retry.
- Suppose there is something wrong with the node causing it to stay in unstarted state forever. If we see we have an unstarted node  try to unregister and register immediately we will always have a node in an unstarted state. As a result no other node will try to register and our deployment will remain broken. If we give a time window for other nodes to enter the cluster the admin will be able to identify that a certain node is failing and take it down
- I am not sure how long we should wait before we try to register again.

Both of the above points are speculation be cause again this bug is so rare and i cannot reproduce. I would be ok if we decide to not merge this but on the other hand the cluster is already broken at this point so we might as well try to recover be evicting the failing node.